p2p/peerinfo/peerinfo.go

```diff
@@ -0,0 +1,389 @@
+
+const (
+	KindUknown            Kind = ""
+	KindInbound           Kind = "inbound"
+	KindOutbound          Kind = "outbound"
+	KindHolePunchInbound  Kind = "hp_inbound"
+	KindHolePunchOutbound Kind = "hp_outbound"
+	KindHolePunchUnknown  Kind = "hp_unknown"
+	KindRelayInbound      Kind = "relay_in"
+	KindRelayOutbound     Kind = "relay_out"
+
+	bpsInterval1 = 10 * time.Second
+	bpsInterval2 = 5 * time.Minute
+
+	totalProto = "__total__"
+	otherProto = "__other__"
+)
+
+type PeerRequestStats struct {
+	sync.Mutex
+	successCount int
+	failureCount int
+	duration     time.Duration
+}
+
+func (ps *PeerRequestStats) SuccessCount() int {
+	ps.Lock()
+	defer ps.Unlock()
+	return ps.successCount
+}
+
+func (ps *PeerRequestStats) FailureCount() int {
+	ps.Lock()
+	defer ps.Unlock()
+	return ps.failureCount
+}
+
+func (ps *PeerRequestStats) Latency() time.Duration {
+	ps.Lock()
+	defer ps.Unlock()
+	count := ps.successCount + ps.failureCount
+	if count == 0 {
+		return 0
+	}
+	return ps.duration / time.Duration(count)
+}
+
+func (ps *PeerRequestStats) RequestDone(took time.Duration, success bool) {
+	ps.Lock()
+	defer ps.Unlock()
+	if success {
+		ps.successCount++
+	} else {
+		ps.failureCount++
+	}
+	ps.duration += took
+}
+
+type DataStats struct {
+	sync.Mutex
+	// [0] is the current value
+	// [1] is prev value for rate 1
+	// [2] is prev value for rate 2
+	bytesSent     [3]int64
+	bytesReceived [3]int64
+	// [0] is for the shorter-interval rate, [1] is for the longer-interval rate
+	recvRate [2]int64
+	sendRate [2]int64
+}
+
+func rateIndex(which int) int {
+	switch which {
+	case 1, 2:
+		return which - 1
+	default:
+		panic("bad rate index")
+	}
+}
+
+func bpsInterval(which int) time.Duration {
+	switch which {
+	case 1:
+		return bpsInterval1
+	case 2:
+		return bpsInterval2
+	default:
+		panic("bad rate index")
+	}
+}
+
+func (ds *DataStats) Tick(which int) {
+	ds.Lock()
+	defer ds.Unlock()
+	interval := bpsInterval(which)
+	idx := rateIndex(which)
+	ds.recvRate[idx] = int64(float64(ds.bytesReceived[0]-ds.bytesReceived[which]) / interval.Seconds())
+	ds.sendRate[idx] = int64(float64(ds.bytesSent[0]-ds.bytesSent[which]) / interval.Seconds())
+	ds.bytesSent[which] = ds.bytesSent[0]
+	ds.bytesReceived[which] = ds.bytesReceived[0]
+}
+
+func (ds *DataStats) RecvRate(which int) int64 {
+	ds.Lock()
+	defer ds.Unlock()
+	return ds.recvRate[rateIndex(which)]
+}
+
+func (ds *DataStats) SendRate(which int) int64 {
+	ds.Lock()
+	defer ds.Unlock()
+	return ds.sendRate[rateIndex(which)]
+}
+
+func (ds *DataStats) RecordSent(n int64) {
+	ds.Lock()
+	defer ds.Unlock()
+	ds.bytesSent[0] += n
+}
+
+func (ds *DataStats) RecordReceived(n int64) {
+	ds.Lock()
+	defer ds.Unlock()
+	ds.bytesReceived[0] += n
+}
+
+func (ds *DataStats) BytesSent() int64 {
+	ds.Lock()
+	defer ds.Unlock()
+	return ds.bytesSent[0]
+}
+
+func (ds *DataStats) BytesReceived() int64 {
+	ds.Lock()
+	defer ds.Unlock()
+	return ds.bytesReceived[0]
+}
+
+type Info struct {
+	DataStats
```

Comment:
should this be public?

Reply:
Yes, that provides data transfer stats for the peer without per-protocol breakdown